panic: tomb.Go called after all goroutines terminated

panic: tomb.Go called after all goroutines terminated

goroutine 37 [running]:
runtime.panic(0x859700, 0xc20c0e9e30)
        /usr/lib/go/src/pkg/runtime/panic.c:279 +0xf5
gopkg.in/tomb%2ev2.(*Tomb).Go(0xc209d50db0, 0xc209d50ea0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:145 +0xe8
gopkg.in/hockeypuck/conflux.v2/recon.(*Peer).handleConfig(0xc208026460, 0x7ff44a41ef20, 0xc20cd48560, 0xa17c10, 0x6, 0xa01310, 0x0, 0x0, 0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/hockeypuck/conflux.v2/recon/peer.go:374 +0x32e
gopkg.in/hockeypuck/conflux.v2/recon.(*Peer).InitiateRecon(0xc208026460, 0x7ff44a41eec0, 0xc209d50a80, 0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/hockeypuck/conflux.v2/recon/gossip.go:113 +0x2b3
gopkg.in/hockeypuck/conflux.v2/recon.(*Peer).Gossip(0xc208026460, 0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/hockeypuck/conflux.v2/recon/gossip.go:59 +0x5cb
gopkg.in/hockeypuck/conflux.v2/recon.*Peer.GossipÂ·fm(0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/hockeypuck/conflux.v2/recon/peer.go:133 +0x38
gopkg.in/tomb%2ev2.(*Tomb).run(0xc208026488, 0xc208945840)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:153 +0x23
created by gopkg.in/tomb%2ev2.(*Tomb).Go
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:149 +0x110

goroutine 16 [chan receive, 2643 minutes]:
gopkg.in/tomb%2ev2.(*Tomb).Wait(0xc2080400d0, 0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:116 +0x5d
github.com/hockeypuck/server.(*Server).Wait(0xc208040090, 0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/hockeypuck/server/server.go:285 +0x45
main.main()
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/hockeypuck/server/cmd/hockeypuck/main.go:68 +0x46f

goroutine 19 [finalizer wait, 2643 minutes]:
runtime.park(0x413ce0, 0xd543d0, 0xd51de9)
        /usr/lib/go/src/pkg/runtime/proc.c:1369 +0x89
runtime.parkunlock(0xd543d0, 0xd51de9)
        /usr/lib/go/src/pkg/runtime/proc.c:1385 +0x3b
runfinq()
        /usr/lib/go/src/pkg/runtime/mgc0.c:2644 +0xcf
runtime.goexit()
        /usr/lib/go/src/pkg/runtime/proc.c:1445

goroutine 20 [syscall, 2643 minutes]:
os/signal.loop()
        /usr/lib/go/src/pkg/os/signal/signal_unix.go:21 +0x1e
created by os/signal.initÂ·1
        /usr/lib/go/src/pkg/os/signal/signal_unix.go:27 +0x32

goroutine 21 [chan receive, 2643 minutes]:
database/sql.(*DB).connectionOpener(0xc208046280)
        /usr/lib/go/src/pkg/database/sql/sql.go:583 +0x48
created by database/sql.Open
        /usr/lib/go/src/pkg/database/sql/sql.go:442 +0x27c

goroutine 17 [syscall, 2643 minutes]:
runtime.goexit()
        /usr/lib/go/src/pkg/runtime/proc.c:1445

goroutine 22 [select]:
github.com/syndtr/goleveldb/leveldb/util.(*BufferPool).drain(0xc2080261c0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/syndtr/goleveldb/leveldb/util/buffer_pool.go:205 +0x1d0
created by github.com/syndtr/goleveldb/leveldb/util.NewBufferPool
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/syndtr/goleveldb/leveldb/util/buffer_pool.go:236 +0x202

goroutine 23 [select, 178 minutes]:
github.com/syndtr/goleveldb/leveldb.(*DB).compactionError(0xc20809a580)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/syndtr/goleveldb/leveldb/db_compaction.go:120 +0x3df
created by github.com/syndtr/goleveldb/leveldb.openDB
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/syndtr/goleveldb/leveldb/db.go:126 +0x518

goroutine 24 [select]:
github.com/syndtr/goleveldb/leveldb.(*DB).mpoolDrain(0xc20809a580)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/syndtr/goleveldb/leveldb/db_state.go:73 +0xfb
created by github.com/syndtr/goleveldb/leveldb.openDB
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/syndtr/goleveldb/leveldb/db.go:127 +0x530

goroutine 25 [select, 178 minutes]:
github.com/syndtr/goleveldb/leveldb.(*DB).tCompaction(0xc20809a580)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/syndtr/goleveldb/leveldb/db_compaction.go:812 +0x774
created by github.com/syndtr/goleveldb/leveldb.openDB
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/syndtr/goleveldb/leveldb/db.go:130 +0x572

goroutine 26 [select, 242 minutes]:
github.com/syndtr/goleveldb/leveldb.(*DB).mCompaction(0xc20809a580)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/syndtr/goleveldb/leveldb/db_compaction.go:759 +0x235
created by github.com/syndtr/goleveldb/leveldb.openDB
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/syndtr/goleveldb/leveldb/db.go:131 +0x58a

goroutine 27 [select, 2643 minutes]:
github.com/syndtr/goleveldb/leveldb.(*DB).jWriter(0xc20809a580)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/syndtr/goleveldb/leveldb/db_write.go:37 +0x143
created by github.com/syndtr/goleveldb/leveldb.openDB
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/syndtr/goleveldb/leveldb/db.go:132 +0x5a2

goroutine 39 [chan receive, 2643 minutes]:
github.com/hockeypuck/server.funcÂ·007(0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/hockeypuck/server/server.go:326 +0x70
gopkg.in/tomb%2ev2.(*Tomb).run(0xc2080400d0, 0xc20893f9e0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:153 +0x23
created by gopkg.in/tomb%2ev2.(*Tomb).Go
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:149 +0x110

goroutine 33 [IO wait, 7 minutes]:
net.runtime_pollWait(0x7ff44a41fe40, 0x72, 0x0)
        /usr/lib/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc20802b790, 0x72, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc20802b790, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).accept(0xc20802b730, 0xb477e0, 0x0, 0x7ff44a41e4c8, 0xb)
        /usr/lib/go/src/pkg/net/fd_unix.go:419 +0x343
net.(*TCPListener).AcceptTCP(0xc20803abf0, 0x7ff442378fa0, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/tcpsock_posix.go:234 +0x5d
github.com/hockeypuck/server.tcpKeepAliveListener.Accept(0xc20803abf0, 0x0, 0x0, 0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/hockeypuck/server/server.go:308 +0x4e
net/http.(*Server).Serve(0xc208005e60, 0x7ff44a4214f8, 0xc20803abf0, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/http/server.go:1698 +0x91
net/http.Serve(0x7ff44a4214f8, 0xc20803abf0, 0x7ff44a421530, 0xc208093640, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/http/server.go:1576 +0x7c
github.com/hockeypuck/server.(*Server).listenAndServeHKP(0xc208040090, 0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/hockeypuck/server/server.go:338 +0x137
github.com/hockeypuck/server.*Server.(github.com/hockeypuck/server.listenAndServeHKP)Â·fm(0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/hockeypuck/server/server.go:233 +0x38
gopkg.in/tomb%2ev2.(*Tomb).run(0xc2080400d0, 0xc208945800)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:153 +0x23
created by gopkg.in/tomb%2ev2.(*Tomb).Go
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:149 +0x110

goroutine 34 [IO wait]:
net.runtime_pollWait(0x7ff44a41fad0, 0x72, 0x0)
        /usr/lib/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc210635aa0, 0x72, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc210635aa0, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).Read(0xc210635a40, 0xc210565d35, 0x16b9a, 0x360cb, 0x0, 0x7ff44a41e4c8, 0xb)
        /usr/lib/go/src/pkg/net/fd_unix.go:242 +0x34c
net.(*conn).Read(0xc20ae51d08, 0xc210565d35, 0x16b9a, 0x360cb, 0x0, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/net.go:122 +0xe7
net/http.noteEOFReader.Read(0x7ff44a41ffa0, 0xc20ae51d08, 0xc20932d9c8, 0xc210565d35, 0x16b9a, 0x360cb, 0x4b8, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/http/transport.go:1203 +0x72
net/http.(*noteEOFReader).Read(0xc20a8fbaa0, 0xc210565d35, 0x16b9a, 0x360cb, 0x4b8, 0x0, 0x0)
        <autogenerated>:124 +0xca
bufio.(*Reader).Read(0xc20ca10ba0, 0xc210565d35, 0x16b9a, 0x360cb, 0x16b9a, 0x0, 0x0)
        /usr/lib/go/src/pkg/bufio/bufio.go:165 +0x13b
io.(*LimitedReader).Read(0xc20f5060a0, 0xc210565d35, 0x16b9a, 0x360cb, 0xc20981506c, 0x0, 0x0)
        /usr/lib/go/src/pkg/io/io.go:399 +0xd0
net/http.(*body).readLocked(0xc209815040, 0xc210565d35, 0x360cb, 0x360cb, 0x48e152, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/http/transfer.go:577 +0x81
net/http.(*body).Read(0xc209815040, 0xc210565d35, 0x360cb, 0x360cb, 0x0, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/http/transfer.go:572 +0x11a
net/http.(*bodyEOFSignal).Read(0xc209815080, 0xc210565d35, 0x360cb, 0x360cb, 0x0, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/http/transport.go:1126 +0x254
bytes.(*Buffer).ReadFrom(0xc208a39420, 0x7ff442277968, 0xc209815080, 0x49d35, 0x0, 0x0)
        /usr/lib/go/src/pkg/bytes/buffer.go:169 +0x20e
io/ioutil.readAll(0x7ff442277968, 0xc209815080, 0x200, 0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/lib/go/src/pkg/io/ioutil/ioutil.go:33 +0x1ae
io/ioutil.ReadAll(0x7ff442277968, 0xc209815080, 0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/lib/go/src/pkg/io/ioutil/ioutil.go:42 +0x6c
gopkg.in/hockeypuck/hkp.v1/sks.(*Peer).requestChunk(0xc208046400, 0xc2090dc240, 0xc20b775140, 0x64, 0x19d8, 0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/hockeypuck/hkp.v1/sks/recon.go:276 +0x708
gopkg.in/hockeypuck/hkp.v1/sks.(*Peer).requestRecovered(0xc208046400, 0xc2090dc240, 0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/hockeypuck/hkp.v1/sks/recon.go:228 +0x136
gopkg.in/hockeypuck/hkp.v1/sks.(*Peer).handleRecovery(0xc208046400, 0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/hockeypuck/hkp.v1/sks/recon.go:211 +0xfd
gopkg.in/hockeypuck/hkp.v1/sks.*Peer.(gopkg.in/hockeypuck/hkp.v1/sks.handleRecovery)Â·fm(0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/hockeypuck/hkp.v1/sks/recon.go:148 +0x38
gopkg.in/tomb%2ev2.(*Tomb).run(0xc208046448, 0xc208945810)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:153 +0x23
created by gopkg.in/tomb%2ev2.(*Tomb).Go
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:149 +0x110

goroutine 35 [select, 3 minutes]:
gopkg.in/hockeypuck/hkp.v1/sks.(*Peer).pruneStats(0xc208046400, 0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/hockeypuck/hkp.v1/sks/recon.go:133 +0x151
gopkg.in/hockeypuck/hkp.v1/sks.*Peer.(gopkg.in/hockeypuck/hkp.v1/sks.pruneStats)Â·fm(0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/hockeypuck/hkp.v1/sks/recon.go:149 +0x38
gopkg.in/tomb%2ev2.(*Tomb).run(0xc208046448, 0xc208945820)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:153 +0x23
created by gopkg.in/tomb%2ev2.(*Tomb).Go
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:149 +0x110

goroutine 36 [IO wait, 21 minutes]:
net.runtime_pollWait(0x7ff44a41fd90, 0x72, 0x0)
        /usr/lib/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc20802b870, 0x72, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc20802b870, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).accept(0xc20802b810, 0xb477e0, 0x0, 0x7ff44a41e4c8, 0xb)
        /usr/lib/go/src/pkg/net/fd_unix.go:419 +0x343
net.(*TCPListener).AcceptTCP(0xc20803ac58, 0xc200000000, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/tcpsock_posix.go:234 +0x5d
net.(*TCPListener).Accept(0xc20803ac58, 0x0, 0x0, 0x0, 0x0)
        /usr/lib/go/src/pkg/net/tcpsock_posix.go:244 +0x4b
gopkg.in/hockeypuck/conflux.v2/recon.(*Peer).Serve(0xc208026460, 0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/hockeypuck/conflux.v2/recon/peer.go:286 +0x40e
gopkg.in/hockeypuck/conflux.v2/recon.*Peer.ServeÂ·fm(0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/hockeypuck/conflux.v2/recon/peer.go:135 +0x38
gopkg.in/tomb%2ev2.(*Tomb).run(0xc208026488, 0xc208945830)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:153 +0x23
created by gopkg.in/tomb%2ev2.(*Tomb).Go
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:149 +0x110

goroutine 38 [chan receive, 2643 minutes]:
main.funcÂ·001()
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/hockeypuck/server/cmd/hockeypuck/main.go:54 +0x8c
created by main.main
        /home/bminton/public_html/programs/hockeypuck/packaging/src/github.com/hockeypuck/server/cmd/hockeypuck/main.go:66 +0x459

goroutine 40 [chan receive, 2643 minutes]:
gopkg.in/hockeypuck/conflux.v2/recon.funcÂ·004(0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/hockeypuck/conflux.v2/recon/peer.go:281 +0x70
gopkg.in/tomb%2ev2.(*Tomb).run(0xc208026488, 0xc208952400)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:153 +0x23
created by gopkg.in/tomb%2ev2.(*Tomb).Go
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:149 +0x110

goroutine 13570782 [select]:
net/http.(*persistConn).readLoop(0xc20932d970)
        /usr/lib/go/src/pkg/net/http/transport.go:868 +0x829
created by net/http.(*Transport).dialConn
        /usr/lib/go/src/pkg/net/http/transport.go:600 +0x93f

goroutine 13570783 [select]:
net/http.(*persistConn).writeLoop(0xc20932d970)
        /usr/lib/go/src/pkg/net/http/transport.go:885 +0x38f
created by net/http.(*Transport).dialConn
        /usr/lib/go/src/pkg/net/http/transport.go:601 +0x957

goroutine 13571384 [semacquire]:
sync.runtime_Semacquire(0xc20c0e9d0c)
        /usr/lib/go/src/pkg/runtime/sema.goc:199 +0x30
sync.(*WaitGroup).Wait(0xc2080264b8)
        /usr/lib/go/src/pkg/sync/waitgroup.go:129 +0x14b
gopkg.in/hockeypuck/conflux.v2/recon.funcÂ·003(0x0, 0x0)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/hockeypuck/conflux.v2/recon/peer.go:216 +0x50
gopkg.in/tomb%2ev2.(*Tomb).run(0xc208026488, 0xc20c0e9b00)
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:153 +0x23
created by gopkg.in/tomb%2ev2.(*Tomb).Go
        /home/bminton/public_html/programs/hockeypuck/packaging/src/gopkg.in/tomb.v2/tomb.go:149 +0x110

--- a/recon/peer.go
+++ b/recon/peer.go
@@ -327,20 +327,18 @@ func (p *Peer) setReadDeadline(conn net.Conn, d time.Duration) {
 	}
 }
 
-func (p *Peer) handleConfig(conn net.Conn, role string, failResp string) (_ *Config, _err error) {
+func (p *Peer) remoteConfig(conn net.Conn, role string, config *Config) (*Config, error) {
+	var remoteConfig *Config
 	w := bufio.NewWriter(conn)
-	p.setReadDeadline(conn, defaultTimeout)
 
-	config, err := p.settings.Config()
-	if err != nil {
-		return nil, errgo.Mask(err)
-	}
-
-	var handshake tomb.Tomb
-	result := make(chan *Config)
-
-	// Send config to server on connect
-	handshake.Go(func() error {
+	ch := make(chan struct{})
+	var t tomb.Tomb
+	t.Go(func() error {
+		<-ch
+		return nil
+	})
+	t.Go(func() error {
+		<-ch
 		p.logFields(role, log.Fields{"config": config}).Debug("writing config")
 		err := WriteMsg(w, config)
 		if err != nil {
@@ -352,33 +350,81 @@ func (p *Peer) handleConfig(conn net.Conn, role string, failResp string) (_ *Con
 		}
 		return nil
 	})
-
-	// Receive remote peer's config
-	handshake.Go(func() error {
-		defer close(result)
-
+	t.Go(func() error {
+		<-ch
 		p.logFields(role, log.Fields{"remoteAddr": conn.RemoteAddr()}).Debug("reading remote config")
 		var msg ReconMsg
-		msg, err = ReadMsg(conn)
+		msg, err := ReadMsg(conn)
 		if err != nil {
 			return errgo.Mask(err)
 		}
 
-		remoteConfig, ok := msg.(*Config)
+		rconf, ok := msg.(*Config)
 		if !ok {
 			return errgo.Newf("expected remote config, got %+v", msg)
 		}
 
-		result <- remoteConfig
+		remoteConfig = rconf
 		return nil
 	})
-
-	remoteConfig, ok := <-result
-	err = handshake.Wait()
+	close(ch)
+	err := t.Wait()
 	if err != nil {
 		return nil, errgo.Mask(err)
-	} else if !ok {
-		return nil, errgo.New("config handshake failed")
+	}
+	return remoteConfig, nil
+}
+
+func (p *Peer) ackConfig(conn net.Conn) error {
+	w := bufio.NewWriter(conn)
+
+	ch := make(chan struct{})
+	var t tomb.Tomb
+	t.Go(func() error {
+		<-ch
+		return nil
+	})
+	t.Go(func() error {
+		<-ch
+		err := WriteString(w, RemoteConfigPassed)
+		if err != nil {
+			return errgo.Mask(err)
+		}
+		err = w.Flush()
+		if err != nil {
+			return errgo.Mask(err)
+		}
+		return nil
+	})
+	t.Go(func() error {
+		remoteConfigStatus, err := ReadString(conn)
+		if err != nil {
+			return errgo.Mask(err)
+		}
+		if remoteConfigStatus != RemoteConfigPassed {
+			reason, err := ReadString(conn)
+			if err != nil {
+				return errgo.WithCausef(err, ErrRemoteRejectedConfig, "remote rejected config")
+			}
+			return errgo.NoteMask(ErrRemoteRejectedConfig, reason)
+		}
+		return nil
+	})
+	close(ch)
+	return t.Wait()
+}
+
+func (p *Peer) handleConfig(conn net.Conn, role string, failResp string) (_ *Config, _err error) {
+	p.setReadDeadline(conn, defaultTimeout)
+
+	config, err := p.settings.Config()
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+
+	remoteConfig, err := p.remoteConfig(conn, role, config)
+	if err != nil {
+		return nil, errgo.Mask(err)
 	}
 
 	p.logFields(role, log.Fields{"remoteConfig": remoteConfig}).Debug()
@@ -399,6 +445,7 @@ func (p *Peer) handleConfig(conn net.Conn, role string, failResp string) (_ *Con
 		}
 	}
 
+	w := bufio.NewWriter(conn)
 	if failResp != "" {
 		err = conn.SetWriteDeadline(time.Now().Add(3 * time.Second))
 		if err != nil {
@@ -421,36 +468,7 @@ func (p *Peer) handleConfig(conn net.Conn, role string, failResp string) (_ *Con
 		return nil, errgo.Newf("cannot peer: %v", failResp)
 	}
 
-	var acknowledge tomb.Tomb
-	acknowledge.Go(func() error {
-		err = WriteString(w, RemoteConfigPassed)
-		if err != nil {
-			return errgo.Mask(err)
-		}
-		err = w.Flush()
-		if err != nil {
-			return errgo.Mask(err)
-		}
-		return nil
-	})
-
-	acknowledge.Go(func() error {
-		remoteConfigStatus, err := ReadString(conn)
-		if err != nil {
-			return errgo.Mask(err)
-		}
-		if remoteConfigStatus != RemoteConfigPassed {
-			reason, err := ReadString(conn)
-			if err != nil {
-				return errgo.WithCausef(err, ErrRemoteRejectedConfig, "remote rejected config")
-			}
-			return errgo.NoteMask(ErrRemoteRejectedConfig, reason)
-		}
-		return nil
-	})
-
-	// Ensure we were able to complete acknowledgement.
-	err = acknowledge.Wait()
+	err = p.ackConfig(conn)
 	if err != nil {
 		return nil, errgo.Mask(err)
 	}

